ffmpeg/lpms_ffmpeg.c

```diff
@@ -955,6 +965,24 @@ static int open_input(input_params *params, struct input_ctx *ctx)
 #undef dd_err
 }
 
+static int lpms_send_packet(struct input_ctx *ictx, AVCodecContext *dec, AVPacket *pkt)
+{
+    int ret = avcodec_send_packet(dec, pkt);
+    if (ret == 0 && dec == ictx->vc) ictx->pkt_diff++; // increase buffer count for video packets
+    return ret;
+}
+
+static int lpms_receive_frame(struct input_ctx *ictx, AVCodecContext *dec, AVFrame *frame)
+{
+    int ret = avcodec_receive_frame(dec, frame);
+    if (dec != ictx->vc) return ret;
+    if (!ret && frame && frame->pts != -1) {
+      ictx->pkt_diff--; // decrease buffer count for non-sentinel video frames
+      if (ictx->flushing) ictx->sentinel_count = 0;
```

Comment:
Curious, when will this condition be true?

Reply:
which one? 

`if (ictx->flushing) ictx->sentinel_count = 0;`
this is used to reset the sentinel_count counter if we are in flushing mode and received a valid frame back.

the idea is to keep increasing sentinel count if we're flushing and not receiving any valid frames back. once it hits SENTINEL_MAX we call it quits. if at any point we get a valid frame back we reset the counter.

Reply:
> this is used to reset the sentinel_count counter if we are in flushing mode and received a valid frame back.
>
> the idea is to keep increasing sentinel count if we're flushing and not receiving any valid frames back. once it hits SENTINEL_MAX we call it quits. 
>

Gotcha, so this allows us to accommodate cases where we might have to flush more than SENTINEL_MAX (eg, to continue draining the decoder), but that's okay as long as we're still receiving valid frames from the decoder?

Just curious, is there ever a case where we hit SENTINEL_MAX? Would that be a problem that the user might want to be notified of? If so, would it be helpful to return an error specifically for this case (from `send_first_pkt`?), rather than trying to "walk backwards" and infer the issue from some other failure?

Reply:
>  so this allows us to accommodate cases where we might have to flush more than SENTINEL_MAX

Yep!

> is there ever a case where we hit SENTINEL_MAX

It will happen if there are no frames left to decode but our `pkt_diff` is not 0 for some reason (like #155). If we choose to return an error here instead of simply ignoring and moving on, we will break how #155 worked before this PR. Or did you have something else in mind? Logging it as a warning on stdout sounds ok, but I don't see us logging anything rn in the Cgo code.

IMO this is not an error but just a fail-proof to stop our decoder going into an infinite loop if there is any mismatch in `pkt_diff`. For example if we used ffmpeg cli or NULL flushing it would happily output mismatched frames without informing the user about it.

Reply:
> IMO this is not an error but just a fail-proof to stop our decoder going into an infinite loop if there is any mismatch in pkt_diff

Yes, this makes sense now - we might still want to return the frames that we've already received, rather than erroring out and discarding the work that's already been done. (I originally thought this case might only be triggered with very low frame counts, maybe if *nothing* had been output, but that doesn't seem to be the case.) Thanks for the explanation!